FreeRTOS/Demo/ARC_NSIM_QEMU_GCC_MW/arc/arc_builtin.h

```diff
@@ -0,0 +1,154 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+   --------------------------------------------- */
+
+/**
+ * @file
+ * @ingroup ARC_HAL_BUILTIN
+ * @brief Header file of builtin and helper functions
+ *
+ * The Metaware toolchain and the GNU toolchain are supported. The details please go to see the file.
+ */
+
+/**
+ * @addtogroup	ARC_HAL_BUILTIN
+ * @{
+ */
+
+#ifndef _ARC_HAL_BUILTIN_H_
+#define _ARC_HAL_BUILTIN_H_
+
+#include "arc.h"
+
+#ifndef __ASSEMBLY__
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @fn uint32_t arc_aux_read(uint32_t aux)
+ * @brief Read auxiliary register
+ *
+ * @param aux auxiliary register address
+ * @return value of auxiliary register
+ */
+Inline uint32_t arc_aux_read(uint32_t aux)
+{
+	uint32_t ret;
+
+	Asm("lr %0, [%1]"
+		: "=r" (ret)
+		: "r" (aux));
+
+	return ret;
+}
+
+/**
+ * @fn void arc_aux_write(uint32_t aux, uint32_t val)
+ * @brief Write auxiliary register
+ *
+ * @param aux auxiliary register address
+ * @param val Value to write
+ */
+Inline void arc_aux_write(uint32_t aux, uint32_t val)
+{
+	Asm(
+		"sr   %0, [%1]	\n"
+		:
+		: "ir" (val), "r" (aux));
+}
+
+/**
+ * @fn void arc_kflag(uint32_t flag)
+ * @brief Call kflag instruction to change status32
+ *
+ * @param flag Flag to set in status32
+ */
+Inline void arc_kflag(uint32_t flag)
+{
+	/*sr cannot write AUX_STATUS32 */
+	Asm("kflag %0" ::"ir" (flag));
+}
+
+/**
+ * @fn uint32_t arc_clri(void)
+ * @brief Call clri instruction
+ * call a clri instruction to disable interrupt
+ * @return interrupt related bits in status32
+ */
+Inline uint32_t arc_clri(void)
+{
+	uint32_t v;
+
+	Asm("clri %0" : "=r" (v):: "memory");
+	return v;
+}
+
+/**
+ * @fn void arc_seti(uint32_t key)
+ * @brief Call seti instruction
+ * call a set instruction to change interrupt status
+ * @param key  interrupt status
+ */
+Inline void arc_seti(uint32_t key)
+{
+	Asm("seti %0" : : "ir" (key) : "memory");
+}
+
+/**
+ * @fn int32_t arc_goto_main(int32_t argc, char **argv)
+ * @brief Go to main function with proper arguments
+ * @param argc Argument count
+ * @param argv Argument content array
+ * @retval Return value of main function
+ */
+Inline int32_t arc_goto_main(int32_t argc, char **argv)
+{
+	int32_t ret;
+
+	Asm(
+		"mov %%r0, %1\n"
+		"mov %%r1, %2\n"
+		"push_s %%blink\n"
+		"jl main\n"
+		"pop_s %%blink\n"
+		"mov %0, %%r0"
+		: "=r" (ret) : "r" (argc), "r" (argv));
+	return (int32_t)ret;
+}
```

Comment:
Why it's in ASM? Why do we even need it?

Reply:
copied from osp, the original purpose is to move to `main()` and pass args. But we don't have args in this demo, now I dropped this function.